src/overlays/actors/ovl_En_Raf/z_en_raf.c

```diff
@@ -80,65 +178,787 @@ static DamageTable D_80A1939C = {
+        this->dyna.actor.flags |= (ACTOR_FLAG_1 | ACTOR_FLAG_4);
+    }
+
+    this->actionFunc = EnRaf_PostDetonation;
+}
+
+/**
+ * Switches the lily pad to the "dead" state once the timer reaches 0.
+ */
+void EnRaf_PostDetonation(EnRaf* this, GlobalContext* globalCtx) {
+    if (this->timer == 0) {
+        this->collider.dim.radius = 50;
+        this->collider.dim.height = 10;
+        func_800BC154(globalCtx, &globalCtx->actorCtx, &this->dyna.actor, 6);
+        this->dyna.actor.flags &= ~(ACTOR_FLAG_1 | ACTOR_FLAG_4);
+        EnRaf_SetupDeadIdle(this);
+    } else if (this->grabTarget == EN_RAF_GRAB_TARGET_EXPLOSIVE) {
+        this->collider.dim.radius = 80;
+        this->collider.dim.height = 50;
+        CollisionCheck_SetAT(globalCtx, &globalCtx->colChkCtx, &this->collider.base);
+    }
+}
+
+void EnRaf_SetupConvulse(EnRaf* this) {
+    EnRaf_ChangeAnimation(this, EN_RAF_ANIMATION_CONVULSE);
+    this->chewCount = 0;
+    this->action = EN_RAF_ACTION_CONVULSE;
+    this->actionFunc = EnRaf_Convulse;
+}
+
+/**
+ * Plays the convulsing animation and sets the lily pad's switch flag to prevent them from
+ * ever coming back to life. When the water in Woodfall Temple is purified, this function
+ * and EnRaf_Dissolve are jointly responsible for controlling the lily pad's death.
+ */
+void EnRaf_Convulse(EnRaf* this, GlobalContext* globalCtx) {
+    f32 curFrame = this->skelAnime.curFrame;
+
+    if (this->endFrame <= curFrame) {
+        this->chewCount++;
+        if (this->chewCount > (BREG(2) + 2)) {
+            if (this->switchFlag >= 0) {
+                Flags_SetSwitch(globalCtx, this->switchFlag);
+            }
+
+            EnRaf_SetupDissolve(this);
+        }
+    }
+}
+
+void EnRaf_SetupDissolve(EnRaf* this) {
+    EnRaf_ChangeAnimation(this, EN_RAF_ANIMATION_DEATH);
+    this->action = EN_RAF_ACTION_DISSOLVE;
+    this->dissolveTimer = 0;
+    this->actionFunc = EnRaf_Dissolve;
+}
+
+/**
+ * Makes the trap petals on the lily pad dissolve and switches the lily pad to the "dead" state.
+ * When the water in Woodfall Temple is purified, this function and EnRaf_Convulse are jointly
+ * responsible for controlling the lily pad's death.
+ */
+void EnRaf_Dissolve(EnRaf* this, GlobalContext* globalCtx) {
+    f32 curFrame = this->skelAnime.curFrame;
+    s32 i;
+
+    if (this->endFrame <= curFrame) {
+        this->dissolveTimer++;
+        if (this->dissolveTimer < (BREG(3) + 105)) {
+            for (i = 0; i < (BREG(4) + 5); i++) {
+                EnRaf_ClearPixelsPetal(Lib_SegmentedToVirtual(&gCarnivorousLilyPadTrapPetalTex),
+                                       sPetalClearPixelTableFirstPass, this->petalClearPixelFirstPassIndex);
+                EnRaf_ClearPixelsTeeth(Lib_SegmentedToVirtual(&gCarnivorousLilyPadTrapTeethTex),
+                                       sTeethClearPixelTableFirstPass, this->teethClearPixelFirstPassIndex);
+                if (this->petalClearPixelFirstPassIndex < 0x200) {
```

Comment:
I suppose this is related to the texture dimensions?

Reply:
Yep, it's 16 * 32. Should I express it as such, like this? I would do this with all texture dimension stuff.
```c
if (this->petalClearPixelFirstPassIndex < (16 * 32))
```

Reply:
Yes, once zeldaret/ZAPD#233 is in we'll have macros for them anyway though.

Reply:
Ah, well, let me just do them manually for now then. If I forget to come back to this once the ZAPD change is merged, feel free to remind me.